language-server/src/main/java/org/sonarlint/languageserver/SonarLintLanguageServer.java

```diff
@@ -220,34 +250,143 @@ void error(String message, Throwable t) {
     telemetry.init(getStoragePath(productKey, telemetryStorage), productName, productVersion);
     telemetry.optOut(userSettings.disableTelemetry);
 
-    info("Starting SonarLint engine...");
+    typeScriptLocation = (String) options.get(TYPESCRIPT_LOCATION);
+
+    startStandaloneEngine();
+
+    initServerInfoCache(options.get(CONNECTED_MODE_SERVERS_PROP));
+    updateBinding((Map<?, ?>) options.get(CONNECTED_MODE_PROJECT_PROP));
+
+    InitializeResult result = new InitializeResult();
+    ServerCapabilities c = new ServerCapabilities();
+    c.setTextDocumentSync(getTextDocumentSyncOptions());
+    c.setCodeActionProvider(true);
+    c.setExecuteCommandProvider(new ExecuteCommandOptions(SONARLINT_COMMANDS));
+    c.setWorkspace(getWorkspaceServerCapabilities());
+
+    result.setCapabilities(c);
+    return CompletableFuture.completedFuture(result);
+  }
+
+  private static WorkspaceServerCapabilities getWorkspaceServerCapabilities() {
+    WorkspaceFoldersOptions options = new WorkspaceFoldersOptions();
+    options.setSupported(true);
+    options.setChangeNotifications(true);
+
+    WorkspaceServerCapabilities capabilities = new WorkspaceServerCapabilities();
+    capabilities.setWorkspaceFolders(options);
+    return capabilities;
+  }
+
+  private static TextDocumentSyncOptions getTextDocumentSyncOptions() {
+    TextDocumentSyncOptions textDocumentSyncOptions = new TextDocumentSyncOptions();
+    textDocumentSyncOptions.setOpenClose(true);
+    textDocumentSyncOptions.setChange(TextDocumentSyncKind.Full);
+    textDocumentSyncOptions.setSave(new SaveOptions(true));
+    return textDocumentSyncOptions;
+  }
+
+  private void initServerInfoCache(@Nullable Object connectedModeServers) {
+    if (connectedModeServers == null) {
+      return;
+    }
+
+    try {
+      List<Map<String, String>> listOfMaps = (List<Map<String, String>>) connectedModeServers;
+      listOfMaps.forEach(m -> {
+        String serverId = m.get("serverId");
+        String url = m.get("serverUrl");
+        String token = m.get("token");
+        String organization = m.get("organizationKey");
+        if (!isBlank(serverId) && !isBlank(url) && !isBlank(token)) {
+          serverInfoCache.put(serverId, new ServerInfo(serverId, url, token, organization));
+        } else {
+          warn("Some required parameters are missing or blank: serverId, url, token");
+        }
+      });
+    } catch (Exception e) {
+      debug("Could not parse the value of " + CONNECTED_MODE_SERVERS_PROP + ": " + connectedModeServers);
+    }
+  }
+
+  private static boolean isBlank(@Nullable String s) {
+    return s == null || s.isEmpty();
+  }
+
+  private void handleUpdateServerStorageCommand(@Nullable List<Object> arguments) {
+    // TODO only update servers that have changed
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=192&issues=AWMr4Qgbau6S9RpGg66y&open=AWMr4Qgbau6S9RpGg66y)

[ISSUE_KEY]: AWMr4Qgbau6S9RpGg66y